actor/root_context.go

```diff
@@ -0,0 +1,116 @@
+package actor
+
+import "time"
+
+type RootContext struct {
+	senderMiddleware SenderFunc
+	headers          messageHeader
+}
+
+var emptyRootContext = &RootContext{
+	senderMiddleware: nil,
+	headers:          emptyMessageHeader,
+}
+
+func EmptyRootContext() *RootContext {
+	return emptyRootContext
+}
+
+func NewRootContext(header map[string]string, middleware ...SenderMiddleware) *RootContext {
+	return &RootContext{
+		senderMiddleware: makeSenderMiddlewareChain(middleware, func(_ SenderContext, target *PID, envelope *MessageEnvelope) {
+			target.sendUserMessage(envelope)
+		}),
+		headers: messageHeader(header),
+	}
+}
+
+func (rc *RootContext) WithHeaders(headers map[string]string) *RootContext {
+	rc.headers = headers
+	return rc
+}
+
+func (rc *RootContext) WithSenderMiddleware(middleware ...SenderMiddleware) *RootContext {
+	rc.senderMiddleware = makeSenderMiddlewareChain(middleware, func(_ SenderContext, target *PID, envelope *MessageEnvelope) {
+		target.sendUserMessage(envelope)
+	})
+	return rc
+}
+
+//
+// Interface: SenderContext
+//
+
+func (rc *RootContext) Message() interface{} {
+	return nil
+}
+
+func (rc *RootContext) MessageHeader() ReadonlyMessageHeader {
+	return rc.headers
+}
+
+func (rc *RootContext) Send(pid *PID, message interface{}) {
+	rc.sendUserMessage(pid, message)
+}
+
+func (rc *RootContext) Request(pid *PID, message interface{}) {
+	rc.sendUserMessage(pid, message)
+}
+
+func (rc *RootContext) RequestWithCustomSender(pid *PID, message interface{}, sender *PID) {
+	env := &MessageEnvelope{
+		Header:  nil,
+		Message: message,
+		Sender:  sender,
+	}
+	rc.sendUserMessage(pid, env)
+}
+
+// RequestFuture sends a message to a given PID and returns a Future
+func (rc *RootContext) RequestFuture(pid *PID, message interface{}, timeout time.Duration) *Future {
+	future := NewFuture(timeout)
+	env := &MessageEnvelope{
+		Header:  nil,
+		Message: message,
+		Sender:  future.PID(),
+	}
+	rc.sendUserMessage(pid, env)
+	return future
+}
+
+func (rc *RootContext) sendUserMessage(pid *PID, message interface{}) {
+	if rc.senderMiddleware != nil {
+		if envelope, ok := message.(*MessageEnvelope); ok {
+			// Request based middleware
+			rc.senderMiddleware(rc, pid, envelope)
+		} else {
+			// tell based middleware
+			rc.senderMiddleware(rc, pid, &MessageEnvelope{nil, message, nil})
```

Comment:
Shouldn't this be this?:

```suggestion
			rc.senderMiddleware(rc, pid, &MessageEnvelope{rc.headers, message, nil})
```